Multiplex: Make RequestId generic

koivunej opened the pull request:
Fixes #129 

Summary of changes:

 * add trait `streaming::multiplex::RequestIdSource` for generating `RequestId`s
   * trait can either take the id out of the message or generate it, like `Counter`
   * refactored old incrementing `u64` counter as  `streaming::multiplex::Counter`
   * created in `streaming::multiplex::ClientProto`, typename is `RequestIds`
 * add trait `streaming::multiplex::RequestId` which is `Hash + Eq + Copy + Debug`
 * all traits and structs that depended on `RequestId` have now received associated type `RequestId`
 * tests now use `u64` explicitly as `Frame<u64, ...>`
 * `simple::multiplex::*` continues to use `u64` and `streaming::multiplex::Counter`

### Naming 

Originally I was thinking about naming the new requirements for `RequestId` as `RId` and to keep the old for backwards compatibility. In the latest commit `RId` has been removed in favor of recycling the old `RequestId` name.

The associated type for `RequestIdSouce` in `streaming::multiplex::client::ClientProto` is `RequestIds`. This could be more descriptive, open for suggestions.

### `simple::multiplex::*` using `u64` and `Counter`

I am not sure if this makes much sense, and if there are no objections I'll try to make these generic as well. This would keep the *simple* side usable even if streaming is not needed.

----

Edited to reflect latest changes (mostly naming).

- [x] Destiny of old `RequestId` and the new `RequestId`
- [x] Name of `streaming::multiplex::client::ClientProto::RequestIds` (`RequestIdSource`)
- [x] Make `simple::multiplex::*` generic as well

koivunej commented:
Looks like I definetly missed some RequestId usage, trying to fix.

koivunej commented:
Turns out that having the trait named `RequestId` and assoc types named `RequestId` does not cause any conflicts. Since assoc types are always used as `P::RequestId` and such, maybe it's better to just remove recycle the name for the new trait and use `RequestId` everywhere instead of `RId` and `RID` without compromising readability.

koivunej commented on src/streaming/multiplex/mod.rs:
```diff
@@ -17,11 +19,40 @@ pub use self::server::ServerProto;
+/// Uniqueness needs depend on the application and can be wider than single session.
+///
+/// Depending on the protocol the identifier can be generated or embedded in the message `T`.
+pub trait RequestIdSource<Id, T>: 'static {
```
Currently implementing `tokio_proto::multiplex::ClientProto` requires user to also 

```
use tokio_proto::streaming::multiplex::RequestIdSource;
```

which seems quite ackward. Should this be moved or duplicated, or ...?

vorner commented on src/simple/multiplex/client.rs:
```diff
@@ -28,20 +28,30 @@ pub trait ClientProto<T: 'static>: 'static {
     type BindTransport: IntoFuture<Item = Self::Transport, Error = io::Error>;
 
+    /// The `RequestIdSource` to use.
+    type RequestIds: RequestIdSource<Self::RequestId, Self::Request>;
```
Nitpicking, but if it's `RequestIdSource`, maybe the last `s` should be capital?

koivunej replied:
I think that `RequestIds` could have a better name altogether, but haven't figured out a better one yet. One possibility is just using `RequestIdSource` which I tried to initially avoid as I thought it would be confusing, but seems to work with `RequestId` assoc type matching the trait name.

koivunej replied:
Decided to rename this to `RequestIdSource`, perhaps it's more like the other names this way.

vorner commented on src/streaming/multiplex/frame.rs:
```diff
@@ -1,12 +1,10 @@
     Message {
         /// Message exchange identifier
-        id: RequestId,
+        id: RId,
```
Didn't you say that you decided to recycle the RequestId name in the end?

vorner commented on src/streaming/multiplex/mod.rs:
```diff
@@ -32,7 +63,7 @@ pub struct StreamingMultiplex<B>(B);
 ///
 /// All methods added in this trait have default implementations.
-pub trait Transport<ReadBody>: 'static +
+pub trait Transport<RID, ReadBody>: 'static +
```
Again, didn't you want to use `RequestId`? Also, this seems to be inconsistent capitalization (the other one was `RId`).

vorner commented:
I wanted to have a look, but I know very little about the code, so I was able to spot only few style issues. Please, don't consider this a review.

koivunej commented:
Thanks @vorner for taking a look! Latest push should fix the remaining `RI[dD]` names to `RequestId`. Also included is use of `Copy` from Frame replaced in favor of `Clone` with `Frame` API change. Did not do anything for the `RequestIds` yet though, hoping for some more input.

koivunej commented:
Stable osx build failed because of network error while downloading dependencies, others succeeded. I guess there is no reason to retry before next changes.

carllerche commented:
cc @aturon

carllerche commented:
Thanks for the PR (and sorry for the delay in reviewing).

It's definitely helpful to see how this could be implemented. I am worried about a) this being a breaking change (so it could not be merged as part of 0.1). b) that this increases the level of complexity for those who do not require a generic request ID.

I'll try to add some inline comments.

carllerche commented on src/simple/multiplex/client.rs:
```diff
@@ -28,20 +28,30 @@ pub trait ClientProto<T: 'static>: 'static {
     type Response: 'static;
 
+    /// The type of request ids to used to correlate requests to responses
+    type RequestId: RequestId;
```
This would be a breaking change. Also, without the ability to have default associated types, this would increase the burden to get started.

koivunej replied:
With the the change you suggested and assuming [I understood it correctly](https://github.com/tokio-rs/tokio-proto/pull/148#discussion_r105286531) the changes for `u64` tokio-line should now be just a single added line, unless `RequestId` alias was used explicitly like in the examples.

I just pushed [a branch with necessary changes](https://github.com/tokio-rs/tokio-line/compare/master...koivunej:track-generic-requestid) to check that it so it really is just the assoc type.

carllerche commented on src/simple/multiplex/client.rs:
```diff
@@ -28,20 +28,30 @@ pub trait ClientProto<T: 'static>: 'static {
     type BindTransport: IntoFuture<Item = Self::Transport, Error = io::Error>;
 
+    /// The `RequestIdSource` to use.
+    type RequestIdSource: RequestIdSource<Self::RequestId, Self::Request>;
```
I would prefer avoiding a bound here. Maybe have a trait bound on `RequestId` that enables to get the generator.

koivunej replied:
Added a [`NewRequestIdSource`](https://github.com/tokio-rs/tokio-proto/pull/148/commits/e1c40792b0e4c51734000f0121604accc5c7c813#diff-660d3e6ba200ddb014ffe50d25d5e81bR35) trait which is now an additional bound on `RequestId`. Have to admit I didn't think this kind of bound was possible :)

carllerche commented on src/streaming/multiplex/client.rs:
```diff
@@ -31,20 +31,30 @@ pub trait ClientProto<T: 'static>: 'static {
     type ResponseBody: 'static;
 
+    /// The type of request ids to used to correlate requests to responses
+    type RequestId: RequestId;
```
Same as above

koivunej commented:
Thanks for the comments. I'll see if I can remove some bounds this week.

koivunej commented:
Rebased and cleaned bunch of commits (don't want to conflict edit those renames yet again) against the latest master. Tests pass at least locally, travis seems to have some issues. If not resolved in a day I guess I'll fix the dates on the commits to get github display them in commit order..

carllerche commented:
Alright, I know that this PR has been ignored... sorry for that. I just wanted to post an update to say that I haven't forgotten this PR and I like the goals, but it is a big change and I haven't been able to schedule the time to focus on it.

Right now, I've been prioritizing my attention for other (lower level) bits of Tokio. However, I do have plans to switch back to tokio-proto soon(ish) and I will spend time on this PR.

Hopefully that is ok :)

koivunej commented:
Thanks for the update @carllerche! I am still looking forward to getting this merged in one way or another. While this change was a bit complicated for a newbie to do it has been working flawlessly in my main project where the protocol uses `Uuid` as the request id. Getting the lower level stuff done well is of course higher priority and thanks for your continued work on tokio!

Kerollmops commented:
Something new ? 😄

koivunej commented:
@Kerollmops nothing has changed for some time now -- the commits are in non-obvious order as the middle one shown by github is the latest one, thus only it has been tested by CI.

if you want to try this out, you can use my repository as your `tokio-proto` dependency.

carllerche commented:
Sorry for the delay.

As a result of the Tokio RFC, tokio-proto is going to be significantly simplified. As such, a generic RequestId is out of scope. I'm going close this PR.

Thanks again for the work!
